x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java

```diff
@@ -0,0 +1,148 @@
+            return;
+        }
+        List<X509Certificate> caCertificates;
+        try {
+            caCertificates = ((StoreKeyConfig) keyConfig).x509Certificates(environment)
+                .stream()
+                .filter(x509Certificate -> x509Certificate.getBasicConstraints() != -1)
+                .collect(Collectors.toList());
+        } catch (Exception e) {
+            listener.onFailure(new ElasticsearchException("Unable to enroll kibana instance. Cannot retrieve CA certificate " +
+                "for the HTTP layer of the Elasticsearch node.", e));
+            return;
+        }
+        if (caCertificates.size() != 1) {
+            listener.onFailure(new ElasticsearchException(
+                "Unable to enroll kibana instance. Elasticsearch node HTTP layer SSL configuration Keystore " +
+                "[xpack.security.http.ssl.keystore] doesn't contain a single PrivateKey entry where the associated " +
+                "certificate is a CA certificate"));
+        } else {
+            String httpCa;
+            try {
+                httpCa = Base64.getUrlEncoder().encodeToString(caCertificates.get(0).getEncoded());
+            } catch (CertificateEncodingException cee) {
+                listener.onFailure(new ElasticsearchException(
+                    "Unable to enroll kibana instance. Elasticsearch node HTTP layer SSL configuration uses a malformed CA certificate",
+                    cee));
+                return;
+            }
+            final NodesInfoRequest nodesInfoRequest = new NodesInfoRequest().addMetric(NodesInfoRequest.Metric.HTTP.metricName());
+            client.execute(NodesInfoAction.INSTANCE, nodesInfoRequest, ActionListener.wrap(nodesInfoResponse -> {
+                final List<String> nodeList = new ArrayList<>();
+                for (NodeInfo nodeInfo : nodesInfoResponse.getNodes()) {
+                    nodeList.add(nodeInfo.getInfo(HttpInfo.class).getAddress().publishAddress().toString());
+                }
+                // Should we guard against setting the password of the kibana_system user multiple times?
+                final char[] password = generateKibanaSystemPassword();
+                final ChangePasswordRequest changePasswordRequest =
+                    new ChangePasswordRequestBuilder(client).username("kibana_system")
+                        .password(password, Hasher.resolve(XPackSettings.PASSWORD_HASHING_ALGORITHM.get(settings)))
+                        .request();
+                try {
+                    client.execute(ChangePasswordAction.INSTANCE, changePasswordRequest, ActionListener.wrap(response -> {
+                        logger.debug("Successfully set the password for user [kibana_system] during kibana enrollment");
+                        listener.onResponse(new KibanaEnrollmentResponse(new SecureString(password), httpCa, nodeList));
```

Comment:
@azasypkin @legrego 
What is your opinion / preference regarding getting the CA certificate ( that you already know the fingerprint of) explicitly as a field in the response vs reading it from the TLS handshake ? I thought the former is easier/preferable but this is mostly up to the consumer ( Kibana ), so I'm happy to change as needed

Reply:
> I thought the former is easier/preferable

Yeah, it'd be my preference as well. The logic to extract all necessary data from a single place (response) would certainly be easier to maintain and test.

Reply:
+1 to getting this from the es response instead of parsing it out of the TLS handshake

Reply:
Thanks for weighing in folks

Reply:
Kibana already needs to extract the certificate from the TLS handshake in order to do the authentication using the fingerprint from the enrollment token.
If ES duplicates the certificate in the response you would need to code and test for the case where we return a different certificate than we use in TLS.

Reply:
Did we consider updating our ES client libs to at least handle fingerprint matching during TLS handshake for us? I naively (and probably mistakenly) assumed that client-type-agnostic "pre-flight" things like this one can be done inside our client libs,  but I admit my knowledge about planned changes in ES client libs is very limited.

Kibana connects to Elasticsearch via `elasticsearch-js` client that, as far as I'm aware, doesn't expose the raw connection socket, so we might need to make a separate "pre-flight" handshake with Elasticsearch host outside of `elasticsearch-js` client just for the sake of fingerprint matching. But even in this case I'd still prefer to have a CA certificate in the response body.

Reply:
> Did we consider updating our ES client libs to at least handle fingerprint matching during TLS handshake for us? 

Yes this is the plan. There will be a new setting that one can set the fingerpint as a value for and the client will do the verification on TLS handshake.

Reply:
It's great to know, thanks!

Reply:
After discussing this, we realized that we do not need to return the other node addresses, at least for now. If we ever have need for this to support HA setups, we can revisit the decision